File: backend/app/engine/roofing.py
"""Roof sheeting / cladding. Unit: m2.

Covered area with a lap/overlap allowance (GI/AC/PPGI sheets), openings netted.
Mirrors the TS engine's roofing.roof_sheeting. See project.md section 6.
"""
from __future__ import annotations

from .units import FormulaStep, Quantity, mm_to_m

CLAUSE = "IS 1200 Part 22"


def roof_sheeting(m) -> list[Quantity]:
    L, B = mm_to_m(m.length_mm), mm_to_m(m.breadth_mm)
    n = m.count
    lap = 1 + m.lap_pct / 100.0
    area = max(L * B - m.opening_area_m2, 0.0) * n * lap
    return [Quantity(
        category="roofing",
        description=f"Roof sheeting {m.label} ({m.length_mm:.0f}x{m.breadth_mm:.0f} mm)".strip(),
        unit="m2", value=area, nos=n, length_m=L, breadth_m=B,
        audit=[FormulaStep(
            "roofing.sheeting.area",
            "max(L*B - openings, 0) * count * (1 + lap_pct)",
            {"L_m": L, "B_m": B, "opening_area_m2": m.opening_area_m2,
             "lap_pct": m.lap_pct, "count": n}, area, CLAUSE)],
    )]

File: frontend/src/engine/roofing.ts
// Roof sheeting / cladding. Ported from roofing.py. Unit: m2.
// Covered area with a lap/overlap allowance (GI/AC/PPGI sheets), openings netted.
import { Member } from "./members";
import { Quantity, mmToM, qty, step, fmt0 } from "./units";

const CLAUSE = "IS 1200 Part 22";

export function roof_sheeting(m: Member): Quantity[] {
  const L = mmToM(m.length_mm), B = mmToM(m.breadth_mm);
  const n = m.count;
  const lap = 1 + (m.lap_pct ?? 0) / 100.0;
  const opening = m.opening_area_m2 ?? 0;
  const area = Math.max(L * B - opening, 0.0) * n * lap;
  return [qty({
    category: "roofing",
    description: `Roof sheeting ${m.label} (${fmt0(m.length_mm)}x${fmt0(m.breadth_mm)} mm)`.trim(),
    unit: "m2", value: area, nos: n, length_m: L, breadth_m: B,
    audit: [step("roofing.sheeting.area",
      "max(L*B - openings, 0) * count * (1 + lap_pct)",
      { L_m: L, B_m: B, opening_area_m2: opening, lap_pct: m.lap_pct ?? 0, count: n },
      area, CLAUSE)],
  })];
}
